File: toggly-ruby/toggly/lib/toggly/user_agent_parser.rb
# frozen_string_literal: true

module Toggly
  # Best-effort User-Agent parse result for segment filters.
  class ParsedUserAgent
    # @return [String]
    attr_reader :browser_family, :os_family, :device_family

    def initialize(browser_family:, os_family:, device_family:)
      @browser_family = browser_family
      @os_family = os_family
      @device_family = device_family
    end
  end

  # Best-effort UA parsing (parity with toggly-eval / Java / Python).
  module UserAgentParser
    module_function

    # @param user_agent [String, nil]
    # @return [ParsedUserAgent, nil]
    def parse(user_agent)
      return nil if user_agent.nil? || user_agent.empty?

      ParsedUserAgent.new(
        browser_family: detect_browser(user_agent),
        os_family: detect_os(user_agent),
        device_family: detect_device(user_agent)
      )
    end

    def detect_browser(user_agent)
      return "Edge" if user_agent.include?("Edg/") || user_agent.include?("EdgiOS/")
      return "Opera" if user_agent.include?("OPR/") || user_agent.include?("Opera")
      return "Chrome" if user_agent.include?("Chrome/") || user_agent.include?("CriOS/")
      return "Firefox" if user_agent.include?("Firefox/") || user_agent.include?("FxiOS/")
      if user_agent.include?("Safari/") && user_agent.include?("Version/") &&
         !user_agent.include?("Chrome") && !user_agent.include?("Chromium")
        return "Safari"
      end

      "Other"
    end
    module_function :detect_browser

    def detect_os(user_agent)
      return "Android" if user_agent.include?("Android")
      if user_agent.include?("iPhone") || user_agent.include?("iPad") || user_agent.include?("iPod") ||
         user_agent.include?("CPU iPhone OS") || user_agent.include?("CPU OS")
        return "iOS"
      end
      return "Mac OS" if user_agent.include?("Mac OS X") || user_agent.include?("Macintosh")
      return "Windows" if user_agent.include?("Windows")
      return "Linux" if user_agent.include?("Linux")

      "Other"
    end
    module_function :detect_os

    def detect_device(user_agent)
      return "iPhone" if user_agent.include?("iPhone")
      return "iPad" if user_agent.include?("iPad")
      return "iPod" if user_agent.include?("iPod")

      "Other"
    end
    module_function :detect_device
  end
end

File: Toggly.FeatureManagement.Java/toggly-core/src/main/java/io/toggly/core/eval/UserAgentParser.java
package io.toggly.core.eval;

/**
 * Best-effort User-Agent parser for segment filters.
 *
 * <p>Bit-identical trees vs ua-parser-js are not required; golden fixtures use
 * common Chrome / iPhone Safari strings.</p>
 */
public final class UserAgentParser {

    private UserAgentParser() {}

    public static final class ParsedUserAgent {
        private final String browserFamily;
        private final String osFamily;
        private final String deviceFamily;

        ParsedUserAgent(String browserFamily, String osFamily, String deviceFamily) {
            this.browserFamily = browserFamily;
            this.osFamily = osFamily;
            this.deviceFamily = deviceFamily;
        }

        public String getBrowserFamily() {
            return browserFamily;
        }

        public String getOsFamily() {
            return osFamily;
        }

        public String getDeviceFamily() {
            return deviceFamily;
        }
    }

    /**
     * Parses a User-Agent string.
     *
     * @param userAgent raw UA header
     * @return parsed fields, or null when UA is blank
     */
    public static ParsedUserAgent parse(String userAgent) {
        if (userAgent == null || userAgent.isEmpty()) {
            return null;
        }
        return new ParsedUserAgent(
                detectBrowser(userAgent),
                detectOs(userAgent),
                detectDevice(userAgent));
    }

    private static String detectBrowser(String ua) {
        if (contains(ua, "Edg/") || contains(ua, "EdgiOS/")) {
            return "Edge";
        }
        if (contains(ua, "OPR/") || contains(ua, "Opera")) {
            return "Opera";
        }
        if (contains(ua, "Chrome/") || contains(ua, "CriOS/")) {
            return "Chrome";
        }
        if (contains(ua, "Firefox/") || contains(ua, "FxiOS/")) {
            return "Firefox";
        }
        if (contains(ua, "Safari/") && contains(ua, "Version/")
                && !contains(ua, "Chrome") && !contains(ua, "Chromium")) {
            return "Safari";
        }
        return "Other";
    }

    private static String detectOs(String ua) {
        if (contains(ua, "Android")) {
            return "Android";
        }
        if (contains(ua, "iPhone") || contains(ua, "iPad") || contains(ua, "iPod")
                || contains(ua, "CPU iPhone OS") || contains(ua, "CPU OS")) {
            return "iOS";
        }
        if (contains(ua, "Mac OS X") || contains(ua, "Macintosh")) {
            return "Mac OS";
        }
        if (contains(ua, "Windows")) {
            return "Windows";
        }
        if (contains(ua, "Linux")) {
            return "Linux";
        }
        return "Other";
    }

    private static String detectDevice(String ua) {
        if (contains(ua, "iPhone")) {
            return "iPhone";
        }
        if (contains(ua, "iPad")) {
            return "iPad";
        }
        if (contains(ua, "iPod")) {
            return "iPod";
        }
        return "Other";
    }

    private static boolean contains(String haystack, String needle) {
        return haystack.contains(needle);
    }
}
